Inject `@config "..."` when a `tailwind.config.{js,ts,...}` is detected

RobinMalfait opened the pull request:
This PR injects a `@config "…"` in the CSS file if a JS based config has been found.

We will try to inject the `@config` in a sensible place:
1. Above the very first `@theme`
2. If that doesn't work, below the last `@import`
3. If that doesn't work, at the top of the file (as a last resort)

RobinMalfait committed:
Ensure `@config "…"` is added when `tailwind.config.js` is detected

thecrypticace commented on packages/@tailwindcss-upgrade/src/codemods/migrate-at-config.ts:
```diff
@@ -0,0 +1,73 @@
+    let configPath = configFilePath
+
+    let relativePath = path.relative(path.dirname(sheetPath), configPath)
+    if (relativePath[0] !== '.') {
+      relativePath = `./${relativePath}`
+    }
```
We'll want to test this on Windows. I think this might be kinda broken there?

RobinMalfait replied:
I think we decided that all these globs should be unix paths with `/`. But _maybe_ the `relativePath` itself might be wrong then. Running the tests as-is on Windows right now.

thecrypticace replied:
yeah I meant `path.relative` is wrong. It uses Windows separators on windows.

RobinMalfait committed:
tmp: run on windows

RobinMalfait committed:
update test to use the `dumpFiles`

RobinMalfait committed:
always use unix file paths with `/`

RobinMalfait committed:
update test with more scenarios

RobinMalfait committed:
improve location node

Inject the `@config` in a sensible place:

1. Above the first `@theme`
2. Below the last `@import`
3. At the top of the file

RobinMalfait committed:
add missing known directives

thecrypticace commented on packages/@tailwindcss-upgrade/src/codemods/migrate-at-config.ts:
```diff
@@ -39,7 +38,7 @@ export function migrateAtConfig(
     let configPath = configFilePath
 
-    let relativePath = path.relative(path.dirname(sheetPath), configPath)
+    let relativePath = path.relative(path.dirname(sheetPath), configPath).replaceAll('\\', '/')
```
I don't completely recall but we have a normalizePath function in `@tailwindcss/node` and I _think_ it can handle relative paths. Might be good to use that here.

RobinMalfait committed:
inject `@config` in root file

If you have a setup where you have an index.css file, which in turn
imports 2 new files with Tailwind CSS imports. In that case we want to
inject the `@config` into the root, not in the individual files.

Input:
```
--- ./src/root.4.css ---
/* Inject missing @config due to nested imports with tailwind imports */
@import './root.4/base.css';
@import './root.4/utilities.css';

--- ./src/root.4/base.css ---
@import 'tailwindcss/base';

--- ./src/root.4/utilities.css ---
@import 'tailwindcss/utilities';
```

Output:
```
--- ./src/root.4.css ---
/* Inject missing @config due to nested imports with tailwind imports */
@import './root.4/base.css';
@import './root.4/utilities.css';
@config "../tailwind.config.ts";

--- ./src/root.4/base.css ---
@import 'tailwindcss/theme' layer(theme);
@import 'tailwindcss/preflight' layer(base);

--- ./src/root.4/utilities.css ---
@import 'tailwindcss/utilities' layer(utilities);"
```

RobinMalfait committed:
use `normalizePath`

This is the same setup as in `./packages/@tailwindcss-vite/src/index.ts`

RobinMalfait committed:
inject `@config` if file has direct `@import "tailwindcss"`

If a non-root file contains `@import "tailwindcss"` (without its
individual imports), then we can inject the `@config` in this file as
well.

This makes it more co-located with where you setup Tailwind CSS

philipp-spiess commented on packages/@tailwindcss-upgrade/src/codemods/migrate-at-config.ts:
```diff
@@ -0,0 +1,109 @@
+    // glob.
+    relative = normalizePath(relative)
+
+    // Inject the `@config` in a sensible place
+    // 1. Above the first `@theme`
+    // 2. Below the last `@import`
+    // 3. At the top of the file
```
This is going to be the place for adding the CSS config if we can migrate the JS config. Maybe we can make the "what is injected" here dynamic so we can decide in the JS config migrator what we need in the css?

RobinMalfait replied:
Yep I think once this PR is merged that we can abstract this to a function if we want. Not sure if that's necessary for this PR yet. 

There might be some additional logic where we might want to control the order of `@config`, `@theme`, injected `@variant`s and so on.

philipp-spiess replied:
Haha yep I would need it right away, let's pair on this afterwards 👍

philipp-spiess reviewed (commented):
Good stuff! Some thoughts on how we can reuse this for the CSS we generate from the JS config too. For the relative path rewrite: If you orient yourself on the other places where we do that, it should be good 👍

RobinMalfait committed:
Revert "tmp: run on windows"

This reverts commit d40e2306dc17f570f5138d7ad605a2c2fa29458c.

RobinMalfait committed:
update changelog

philipp-spiess commented on integrations/upgrade/index.test.ts:
```diff
@@ -32,15 +32,16 @@ test(
+
+      --- ./src/input.css ---
+      @import 'tailwindcss';
+      @config "../tailwind.config.js";"
```
```suggestion
      @config "../tailwind.config.js";
```

RobinMalfait committed:
always inject after last `@import`

RobinMalfait committed:
ensure `dumpFiles(…)` has a clean `\n` at the end

If not, some files could look weird:
```
@import "tailwind.css";"
```

Which feels wrong, but the last `"` is coming from the
`toMatchInlineSnapshot()` function.

RobinMalfait committed:
update snapshots to include `\n`
